src/train.jl

```diff
@@ -0,0 +1,167 @@
+function pm_next!(pm, stats::NamedTuple)
+    return ProgressMeter.next!(pm; showvalues=[tuple(s...) for s in pairs(stats)])
+end
+
+"""
+    train(rng::AbstractRNG, at::ADTypes.AbstractADType, vo, θ₀::AbstractVector{T}, re, args...; max_iters::Int=10000, optimiser::Optimisers.AbstractRule=Optimisers.ADAM(), show_progress::Bool=true, callback=nothing)
+
+Iteratively updating the parameters `θ` of the normalizing flow `re(θ)` by calling `grad!` and using the given `optimiser` to compute the steps.
+
+# Arguments
+- `rng::AbstractRNG`: random number generator
+- `at::ADTypes.AbstractADType`: automatic differentiation backend
+- `vo`: variational objective
+- `θ₀::AbstractVector{T}`: initial parameters of the normalizing flow
+- `re`: function that reconstructs the normalizing flow from the flattened parameters
+- `args...`: additional arguments for `vo`
+- `max_iters::Int=10000`: maximum number of iterations
+- `optimiser::Optimisers.AbstractRule=Optimisers.ADAM()`: optimiser to compute the steps
+- `show_progress::Bool=true`: whether to show the progress bar. The default information printed in the progress bar is the iteration number, the loss value, and the gradient norm.
+- `callback=nothing`: callback function that takes the current iteration number, 
+                        the current optimiser state, 
+                        and the current value of the variational objective as input, 
+                        and returns a dictionary of statistics to be displayed in the progress bar
+
+# Returns
+- `θ`: trained parameters of the normalizing flow
+- `opt_stats`: statistics of the optimiser
+- `st`: optimiser state for potential continuation of training
+"""
+function train(
+    rng::AbstractRNG,
+    at::ADTypes.AbstractADType,
+    vo,
+    θ₀::AbstractVector{T},
+    re,
+    args...;
+    max_iters::Int=10000,
+    optimiser::Optimisers.AbstractRule=Optimisers.ADAM(),
+    show_progress::Bool=true,
+    callback=nothing,
+) where {T<:Real}
+    prog = ProgressMeter.Progress(
+        max_iters; desc="Training", barlen=31, showspeed=true, enabled=show_progress
+    )
+    opt_stats = Vector{NamedTuple}(undef, max_iters)
+
+    θ = copy(θ₀)
+    diff_result = DiffResults.GradientResult(θ)
+    # initialise optimiser state
+    st = Optimisers.setup(optimiser, θ)
+
+    time_elapsed = @elapsed for i in 1:max_iters
+        grad!(rng, at, vo, θ, re, diff_result, args...)
+
+        # save stats
+        ls = DiffResults.value(diff_result)
+        g = DiffResults.gradient(diff_result)
+        stat_ = (iteration=i, loss=ls, gradient_norm=norm(g))
+        opt_stats[i] = stat_
+
+        # callback
+        if !isnothing(callback)
+            new_stat = callback(re, opt_stats, i)
+            stat_ = !isnothing(new_stat) ? merge(new_stat, stat_) : stat_
+        end
+
+        # update optimiser state and parameters
+        st, θ = Optimisers.update!(st, θ, DiffResults.gradient(diff_result))
+        pm_next!(prog, stat_)
```

Comment:
```suggestion
        stat = (iteration=i, loss=ls, gradient_norm=norm(g))
        opt_stats[i] = stat

        # callback
        if !isnothing(callback)
            new_stat = callback(re, opt_stats, i)
            stat = !isnothing(new_stat) ? merge(new_stat, stat) : stat
        end

        # update optimiser state and parameters
        st, θ = Optimisers.update!(st, θ, DiffResults.gradient(diff_result))
        pm_next!(prog, stat)
```
or is there a reason why you didn't do this?

Reply:
no, somehow I just put the sub `_` to indicate that `stat_` is just a intermediate value...